### PR TITLE
Add source_channel and source_guild to Webhook

### PR DESCRIPTION
**Pull request by Zomatree**
## Summary

adds two attributes to webhooks

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/webhook.py [hunk 1]
```diff
@@ -45,10 +45,100 @@
...
     'Webhook',
     'WebhookMessage',
+    'PartialChannel',
+    'PartialGuild'
 )
 
 log = logging.getLogger(__name__)
 
+class PartialChannel(Hashable):
+    """Partial channel for webhooks
```
Comment (Rapptz) [inline discord/webhook.py:55]: This needs more meat.

File: discord/webhook.py [hunk 2]
```diff
@@ -45,10 +45,100 @@
...
+
+    .. versionadded:: 2.0
+
+    Attributes
+    -----------
+    id: :class:`int`
+        The channel's ID.
+    name: :class:`str`
+        The channel's name.
```
Comment (Rapptz) [inline discord/webhook.py:64]: This should probably mention it's a partial webhook channel.

File: discord/webhook.py [hunk 3]
```diff
@@ -45,10 +45,100 @@
...
+
+    def __init__(self, *, data):
+        self.id = int(data['id'])
+        self.name = data['name']
+
+    def __repr__(self):
+        return f'<PartialChannel name=\'{self.name}\' id={self.id}>'
+
+class PartialGuild(Hashable):
+    """Partial guild for webhooks
```
Comment (Rapptz) [inline discord/webhook.py:77]: This also needs more meat.

File: discord/webhook.py [hunk 4]
```diff
@@ -45,10 +45,100 @@
...
+    __slots__ = ('id', 'name', 'icon', '_state')
+
+    def __init__(self, *, data, state):
+        self._state = state
+        self.id = int(data['id'])
+        self.name = data['name']
+        self.icon = data['icon']
+
+    def __repr__(self):
+        return f'<PartialGuild name=\'{self.name}\' id={self.id}>'
```
Comment (Rapptz) [inline discord/webhook.py:100]:
```suggestion
        return f'<PartialWebhookGuild name={self.name!r} id={self.id}>'
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
Some renames.

Note that newer code is typehinted, which this is missing. For now that's okay though.

File: discord/webhook.py [hunk 5]
```diff
@@ -45,10 +45,100 @@
...
     'Webhook',
     'WebhookMessage',
+    'PartialWebhookChannel',
+    'PartialWebhookGuild'
 )
 
 log = logging.getLogger(__name__)
 
+class PartialWebhookChannel(Hashable):
+    """Partial channel for webhooks for when they are channel following webhooks for news channels.
```
Comment (Rapptz) [inline discord/webhook.py:55]: This docstring doesn't match the rest of the codebase.

```suggestion
    """Represents a partial channel for webhooks.

    These are typically given for channel follower webhooks.
```

File: discord/webhook.py [hunk 6]
```diff
@@ -45,10 +45,100 @@
...
+    """
+
+    __slots__ = ('id', 'name')
+
+    def __init__(self, *, data):
+        self.id = int(data['id'])
+        self.name = data['name']
+
+    def __repr__(self):
+        return f'<PartialChannel name=\'{self.name}\' id={self.id}>'
```
Comment (Rapptz) [inline discord/webhook.py:74]:
```suggestion
        return f'<PartialWebhookChannel name={self.name!r} id={self.id}>'
```

Missed this.

File: discord/webhook.py [hunk 7]
```diff
@@ -591,10 +681,19 @@ async def foo():
         The default name of the webhook.
     avatar: Optional[:class:`str`]
         The default avatar of the webhook.
+    source_guild: Optional[:class:`PartialGuild`]
+        The guild of the channel that this webhook is following (returned for Channel Follower Webhooks).
```
Comment (Rapptz) [inline discord/webhook.py:685]:
```suggestion
        The guild of the channel that this webhook is following. Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
```

File: discord/webhook.py [hunk 8]
```diff
@@ -591,10 +681,19 @@ async def foo():
...
+
+        .. versionadded:: 2.0
+
+    source_channel: Optional[:class:`PartialChannel`]
+        The channel that this webhook is following (returned for Channel Follower Webhooks).
```
Comment (Rapptz) [inline discord/webhook.py:690]:
```suggestion
        The channel that this webhook is following. Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/webhook.py [hunk 9]
```diff
@@ -616,6 +717,20 @@ def __init__(self, data, *, adapter, state=None):
         else:
             self.user = User(state=state, data=user)
 
+        source_channel = data.get('source_channel')
+        if source_channel:
+            source_channel = PartialWebhookChannel(data=source_channel)
+        else:
+            source_channel = None
```
Comment (NCPlayz) [inline discord/webhook.py:724]: `source_channel` should be `None` here anyway (unless otherwise?).

File: discord/webhook.py [hunk 10]
```diff
@@ -616,6 +717,20 @@ def __init__(self, data, *, adapter, state=None):
...
+        self.source_channel = source_channel
+
+        source_guild = data.get('source_guild')
+        if source_guild:
+            source_guild = PartialWebhookGuild(data=source_guild, state=state)
+        else:
+            source_guild = None
```
Comment (NCPlayz) [inline discord/webhook.py:731]: `source_guild` should be `None` here anyway (unless otherwise?).

File: discord/webhook.py [hunk 11]
```diff
@@ -591,10 +683,19 @@ async def foo():
         The default name of the webhook.
     avatar: Optional[:class:`str`]
         The default avatar of the webhook.
+    source_guild: Optional[:class:`PartialGuild`]
```
Comment (NCPlayz) [inline discord/webhook.py:686]:
```suggestion
    source_guild: Optional[:class:`PartialWebhookGuild`]
```

File: discord/webhook.py [hunk 12]
```diff
@@ -591,10 +683,19 @@ async def foo():
...
+        The guild of the channel that this webhook is following. Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
+
+        .. versionadded:: 2.0
+
+    source_channel: Optional[:class:`PartialChannel`]
```
Comment (NCPlayz) [inline discord/webhook.py:691]:
```suggestion
    source_channel: Optional[:class:`PartialWebhookChannel`]
```

File: discord/webhook.py [hunk 13]
```diff
@@ -591,10 +683,19 @@ async def foo():
...
+        The channel that this webhook is following. Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
```
Comment (NCPlayz) [inline discord/webhook.py:692]:
```suggestion
        The channel that this webhook is following.
        Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
```

File: discord/webhook.py [hunk 14]
```diff
@@ -591,10 +683,19 @@ async def foo():
         The default name of the webhook.
     avatar: Optional[:class:`str`]
         The default avatar of the webhook.
+    source_guild: Optional[:class:`PartialGuild`]
+        The guild of the channel that this webhook is following. Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
```
Comment (NCPlayz) [inline discord/webhook.py:687]:
```suggestion
        The guild of the channel that this webhook is following.
        Only given if :attr:`type` is :attr:`WebhookType.channel_follower`.
```

File: discord/webhook.py [hunk 15]
```diff
@@ -45,10 +45,102 @@
...
+
+    def __init__(self, *, data):
+        self.id = int(data['id'])
+        self.name = data['name']
+
+    def __repr__(self):
+        return f'<PartialWebhookChannel name={self.name!r} id={self.id}>'
+
+class PartialWebhookGuild(Hashable):
+    """Partial guild for webhooks for when they are channel following webhooks for news channels.
```
Comment (NCPlayz) [inline discord/webhook.py:79]:
```suggestion
    """Represents a partial guild for webhooks.

    These are typically given for channel follower webhooks.
```

**Review by NCPlayz — CHANGES_REQUESTED** [review]

**Review by Rapptz — APPROVED** [review]
LGTM.

**Pull request merged**